File: java/03-bootstrap/src/test/java/com/fResult/bootstrap/customers/services/TransactionTemplateCustomerServiceTest.java
package com.fResult.bootstrap.customers.services;

import com.fResult.bootstrap.TransactionTestMixin;
import com.fResult.bootstrap.common.DataSourceUtils;
import java.util.Optional;
import javax.sql.DataSource;
import org.junit.jupiter.api.Test;
import org.springframework.jdbc.datasource.DataSourceTransactionManager;
import org.springframework.jdbc.datasource.embedded.EmbeddedDatabaseBuilder;
import org.springframework.jdbc.datasource.embedded.EmbeddedDatabaseType;
import org.springframework.transaction.support.TransactionTemplate;

class TransactionTemplateCustomerServiceTest extends BaseClass implements TransactionTestMixin {
  private CustomerService customerService;

  public TransactionTemplateCustomerServiceTest() {
    final var dataSource = new EmbeddedDatabaseBuilder().setType(EmbeddedDatabaseType.H2).build();

    Optional.of(dataSource).map(this::buildCustomerService).ifPresent(it -> customerService = it);
  }

  @Override
  @Test
  public void insert() {
    super.insert();
    this.testTransactionalityOfSave(getCustomerService());
  }

  @Override
  public CustomerService getCustomerService() {
    return customerService;
  }

  private CustomerService buildCustomerService(DataSource dataSource) {
    final var initializedDataSource = DataSourceUtils.initializeDdl(dataSource);
    final var txnManager = new DataSourceTransactionManager(initializedDataSource);
    final var transactionTemplate = new TransactionTemplate(txnManager);

    return new TransactionTemplateCustomerService(initializedDataSource, transactionTemplate);
  }
}

File: kotlin/03-bootstrap/src/test/kotlin/com/fResult/bootstrap/customers/services/TransactionTemplateCustomerServiceTest.kt
package com.fResult.bootstrap.customers.services

import com.fResult.bootstrap.TransactionTestMixin
import com.fResult.bootstrap.common.DataSourceUtils
import org.junit.jupiter.api.Test
import org.springframework.jdbc.datasource.DataSourceTransactionManager
import org.springframework.jdbc.datasource.embedded.EmbeddedDatabaseBuilder
import org.springframework.jdbc.datasource.embedded.EmbeddedDatabaseType
import org.springframework.transaction.support.TransactionTemplate

class TransactionTemplateCustomerServiceTest : BaseClass(), TransactionTestMixin {

  private lateinit var customerService: CustomerService

  @Test
  override fun `should insert customer`() {
    super.`should insert customer`()
    testTransactionalityOfSave(getCustomerService())
  }

  override fun getCustomerService(): CustomerService = EmbeddedDatabaseBuilder()
    .setType(EmbeddedDatabaseType.H2)
    .build()
    .let {
      val dataSource = DataSourceUtils.initializeDdl(it)
      val txnManager = DataSourceTransactionManager(dataSource)
      val transactionTemplate = TransactionTemplate(txnManager)

      TransactionTemplateCustomerService(dataSource, transactionTemplate)
    }.also { customerService = it }
}
